src/ClusterBootstrap/cloud_init_aztools.py

```diff
@@ -546,17 +594,112 @@ def get_deployed_cluster_info(config, args):
+
+def delete_k8s_nodes(node_list, args):
+    for node in node_list:
+        cmd = './ctl.py kubectl delete node {}'.format(node)
+        execute_or_dump_locally(cmd, args.verbose, args.dryrun, args.output)
+
+
+def cordon_node_2_delete_later(num_of_worker_2_cordon, args):
+    if num_of_worker_2_cordon <= 0:
+        return
+    query_cmd = "./ctl.py kubectl get nodes -l worker=active --no-headers | grep Ready | awk '{print $1}'"
+    ready_nodes = get_k8s_node_list_under_condition(query_cmd)
+    for node in ready_nodes[:num_of_worker_2_cordon]:
+        cmd = './ctl.py kubectl cordon {}'.format(node)
+        execute_or_dump_locally(cmd, args.verbose, args.dryrun, args.output)
+
+
+def delete_specified_or_cordoned_idling_nodes(config, args, num_limit=-1):
+    if args.node_list:
+        nodes2delete = args.node_list
+    else:
+        busy_cmd = "./ctl.py kubectl get pods -l type=job -o jsonpath='{.items[*].spec.nodeName}'"
+        busy_nodes = get_k8s_node_list_under_condition(busy_cmd)
+        cordoned_cmd = "./ctl.py kubectl get nodes -l worker=active --no-headers | grep SchedulingDisabled | awk '{print $1}'"
+        # would be [] if no such node
+        cordoned_nodes = get_k8s_node_list_under_condition(cordoned_cmd)
+        cordoned_idling = set(cordoned_nodes) - set(busy_nodes)
+        nodes2delete = cordoned_idling
+        if args.verbose:
+            print("Node list not specified, would delete cordoned idling worker nodes by default")
+    # with :num_limit, we always only delete min(num_limit, # of qualified node) nodes
+    nodes2delete = list(nodes2delete)[:num_limit]
+    if args.verbose:
+        print("Deleting following nodes:\n", nodes2delete)
+    delete_az_vms(config, args, nodes2delete)
+    delete_k8s_nodes(nodes2delete, args)
+    # if we don't have enough qualified node to delete, we randomly cordon some nodes so they'll be cordoned when pod on them finished
+    num_of_worker_2_cordon = num_limit - len(nodes2delete)
+    cordon_node_2_delete_later(num_of_worker_2_cordon, args)
+    if not args.dryrun:
+        get_deployed_cluster_info(config, args)
+
+
+def dynamically_add_or_delete_around_a_num(config, args):
+    # need some time for the newly added worker to register
+    monitor_again_after = config.get("monitor_again_after", 10)
+    while True:
+        # TODO currently don't keep history of operation here. or name the bash by time?
+        os.system("rm -f {}".format(args.output))
+        config = load_config_based_on_command("dynamic_around")
+        dynamic_worker_num = config.get("dynamic_worker_num", -1)
+        if dynamic_worker_num < 0:
+            print("This round would be skipped. Please specify dynamic_worker_num in config.")
+            os.system("sleep {}m".format(monitor_again_after))
+            continue
+        query_cmd = "./ctl.py kubectl get nodes -l worker=active --no-headers | awk '{print $1}'"
+        k8s_worker_nodes = get_k8s_node_list_under_condition(query_cmd)
```

Comment:
What is the use of `k8s_worker_nodes`? The only usage I can think of is to check the consistency between `k8s_worker_nodes` and `worker_in_records`.

Reply:
yes, as shown in the printline, in case some nodes are already created and backed up, but k8s on them not up yet/went down.